src/rgw/driver/d4n/rgw_sal_d4n.cc

```diff
@@ -633,7 +633,11 @@ int D4NFilterObject::set_data_block_dir_entries(const DoutPrefixProvider* dpp, o
     off_t cur_size = std::min<off_t>(fst + dpp->get_cct()->_conf->rgw_max_chunk_size, lst);
     off_t cur_len = cur_size - fst;
     block.cacheObj.bucketName = this->get_bucket()->get_name();
-    block.cacheObj.objName = this->get_key().get_oid();
+    if (this->get_bucket()->versioned() && !this->get_bucket()->versioning_enabled()) {
```

Comment:
why is this check needed?

Reply:
Even here, the version is calculated in writer::prepare() method and used in process() and complete()

Reply:
if the problem is that ReadOp::prepare method does not populate version for the case when versioning is suspended, then in calculate_version method, you need to add the check that if versioning is suspended and then set the instance to 'null' after this line: https://github.com/ceph/ceph/pull/56336/files#diff-5cf52d0baced08513b69bbb3cf4a29682289fce58451b9567c6d2e86ff080bceR516. Something like the following:

```
version = "null";
object->set_instance(version);
```

Reply:
get_instance() returns an empty string when versioning is suspended. For suspended versioning, the block directory will not have "null" in the block name but the block will store "null" as the version. I'll make the suggested change to `calculate_version` and update the rest of the logic